tooling/autodispose-rx-error-prone-checker/src/main/java/com/uber/autodispose/rx/error/prone/checker/RxJavaMissingAutoDisposeErrorChecker.java

```diff
@@ -0,0 +1,197 @@
+import com.google.errorprone.bugpatterns.BugChecker.MethodInvocationTreeMatcher;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.matchers.method.MethodMatchers;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MemberSelectTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.tools.javac.code.Type;
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.errorprone.BugPattern.Category.JDK;
+import static com.google.errorprone.BugPattern.SeverityLevel.ERROR;
+import static com.google.errorprone.matchers.Matchers.instanceMethod;
+
+/**
+ * Checker for subscriptions not binding to lifecycle in components with lifecycle.
+ * Use -XepOpt:AutoDisposeLeakCheck to add support for custom components with lifecycle.
+ */
+@AutoService(BugChecker.class)
+@BugPattern(
+    name = "RxJavaMissingAutoDisposeErrorChecker",
+    summary = "Always apply an Autodispose scope before subscribing",
+    category = JDK,
+    severity = ERROR
+)
+public final class RxJavaMissingAutoDisposeErrorChecker extends BugChecker
+    implements MethodInvocationTreeMatcher {
+
+  private static final String AS = "as";
+  private static final ImmutableList<MethodMatchers.MethodNameMatcher> AS_CALL_MATCHERS;
+  private static final ImmutableList<MethodMatchers.MethodNameMatcher> TO_CALL_MATCHERS;
+  private static final ImmutableList<MethodMatchers.MethodNameMatcher> SUBSCRIBE_MATCHERS;
+  private static final String SUBSCRIBE = "subscribe";
+  private static final String TO = "to";
+
+  private final ImmutableList<String> classesWithLifecycle;
+
+  public RxJavaMissingAutoDisposeErrorChecker(ErrorProneFlags flags) {
+    Optional<ImmutableList<String>> inputClasses = flags.getList("AutoDisposeLeakCheck");
+    ImmutableList.Builder<String> builder = new ImmutableList.Builder<>();
+    classesWithLifecycle = builder
+        .add("android.app.Activity")
+        .add("android.app.Fragment")
+        .add("com.uber.autodispose.LifecycleScopeProvider")
+        .addAll(inputClasses.orElse(ImmutableList.of()))
+        .build();
+  }
+
+  static {
+    ImmutableList.Builder<MethodMatchers.MethodNameMatcher> builder = new ImmutableList.Builder<>();
+    builder
+        .add(instanceMethod().onDescendantOf("io.reactivex.Single").named(TO))
+        .add(instanceMethod().onDescendantOf("io.reactivex.Observable").named(TO))
+        .add(instanceMethod().onDescendantOf("io.reactivex.Completable").named(TO))
+        .add(instanceMethod().onDescendantOf("io.reactivex.Flowable").named(TO))
+        .add(instanceMethod().onDescendantOf("io.reactivex.Maybe").named(TO));
+    TO_CALL_MATCHERS = builder.build();
+
+    ImmutableList.Builder<MethodMatchers.MethodNameMatcher> builder2
+        = new ImmutableList.Builder<>();
+    builder2
```

Comment:
why not continue the fluent API after creating the builder?

Reply:
https://github.com/uber/AutoDispose/pull/156/commits/1b21a9eaef625161af97e5c2b5cdecfc6605df9b